### PR TITLE
Add two new settings `shutdown_wait_extra_unfinished_queries` and `shutdown_wait_extra_unfinished_query_kind` to wait extra for certain kind of queries

### PR DESCRIPTION
**Pull request by chhetripradeep**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add two new settings `shutdown_wait_extra_unfinished_queries` and `shutdown_wait_extra_unfinished_query_kind` in order to make clickhouse wait extra for certain kind of queries

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit c37b7fd3c0a244297d4f67a013bfbe0aaeee546a with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51162/c37b7fd3c0a244297d4f67a013bfbe0aaeee546a/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
</table>

File: programs/server/Server.cpp [hunk 1]
```diff
@@ -1822,9 +1822,21 @@ try
...
+            {
                 global_context->getProcessList().killAllQueries();
+            }
+            /// Killing only selected queries.
+            else
+            {
+                if (!server_settings.shutdown_wait_unfinished_query_kind.value.empty())
+                    for (auto & query : global_context->getProcessList().processes)
+                    {
+                        if (query->getInfo().query_kind != server_settings.shutdown_wait_unfinished_query_type)
```
Comment (qoega) [inline programs/server/Server.cpp:1836]: I don't like that you can set only one query_kind here. Maybe provide some levels?
`1 - insert`, `2 - insert,select`,...

Comment (UnamedRus) [reply programs/server/Server.cpp:1836]: We support multiple values for settings like `join_algorithm='direct, default'`
Why not have here the same?

Comment (chhetripradeep) [reply programs/server/Server.cpp:1836]: Thank you for the reference. I am checking it out.

File: src/Core/SettingsEnums.cpp [hunk 2]
```diff
@@ -42,6 +43,37 @@ IMPLEMENT_SETTING_MULTI_ENUM(JoinAlgorithm, ErrorCodes::UNKNOWN_JOIN,
      {"grace_hash",           JoinAlgorithm::GRACE_HASH}})
 
 
+IMPLEMENT_SETTING_MULTI_ENUM(QueryKind, ErrorCodes::UNKNOWN_QUERY_KIND,
```
Comment (nikitamikhaylov) [inline src/Core/SettingsEnums.cpp:46]: I don't like that we now have two lists of query kinds... I mean this one and `IAST::QueryKind`...

**Review by bianpengyuan — COMMENTED** [review]
I guess we probably want a more granular control here? e.g. for select query, wait for x minutes; but for insert query, wait for y minutes.

**Comment by chhetripradeep** [comment]
> I guess we probably want a more granular control here? e.g. for select query, wait for x minutes; but for insert query, wait for y minutes.

IMHO, waiting for SELECT queries for X minutes and INSERT queries for Y minutes will be same thing as waiting for both SELECT and INSERT queries for Z minutes (where `Z = largest(X, Y)` since in both scenarios clickhouse will be up for Z minutes)

**Comment by serxa** [comment]
It may differ in case we have only one kind of queries at the end. Then Z may be equal to either X or Y. But do we need it?

**Comment by bianpengyuan** [comment]
> IMHO, waiting for SELECT queries for X minutes and INSERT queries for Y minutes will be same thing as waiting for both SELECT and INSERT queries for Z minutes (where Z = largest(X, Y)) since in both scenarios clickhouse will be in terminating state for Z minutes

@chhetripradeep  It is for the case that only SELECT queries are running. I guess we don't want to kill all queries / activities immediately at termination even when there is no insert running?

**Comment by chhetripradeep** [comment]
@bianpengyuan Sorry I misinterpreted your earlier comment. You are right, current implementation isn't the proper way since we are killing rest of the queries immediately.

File: src/Core/SettingsEnums.cpp [hunk 3]
```diff
@@ -43,6 +44,37 @@ IMPLEMENT_SETTING_MULTI_ENUM(JoinAlgorithm, ErrorCodes::UNKNOWN_JOIN,
...
+     {"Describe",               IAST::QueryKind::Describe},
+     {"Explain",                IAST::QueryKind::Explain},
+     {"Backup",                 IAST::QueryKind::Backup},
+     {"Restore",                IAST::QueryKind::Restore},
+     {"KillQuery",              IAST::QueryKind::KillQuery},
+     {"ExternalDDL",            IAST::QueryKind::ExternalDDL},
+     {"Begin",                  IAST::QueryKind::Begin},
+     {"Commit",                 IAST::QueryKind::Commit},
+     {"Rollback",               IAST::QueryKind::Rollback},
+     {"SetTransactionSnapshot", IAST::QueryKind::SetTransactionSnapshot}})
```
Comment (serxa) [inline src/Core/SettingsEnums.cpp:75]:
```suggestion
IMPLEMENT_SETTING_MULTI_AUTO_ENUM(QueryKind, ErrorCodes::UNKNOWN_QUERY_KIND)
```

**Pull request closed without merging**